### PR TITLE
support text embedding task in hugging face server

### LINKED ISSUE #3572: Support text embedding task in huggingface server
/kind feature

**Describe the solution you'd like**

Support text embedding (I think huggingface calls it [Feature Extraction](https://huggingface.co/tasks/feature-extraction)) task in huggingface server. Some examples are [mixedbread-ai/mxbai-embed-large-v1](https://huggingface.co/mixedbread-ai/mxbai-embed-large-v1) and [BAAI/bge-base-en-v1.5](https://huggingface.co/BAAI/bge-base-en-v1.5).

This seems like a pretty common task that people use, mostly to later store these in vector databases, so that they can do nearest neighbor search, or compute similarity between different texts, etc, so I think it'd be a good addition to have.

I have implemented this myself in my fork of kserve, so I'd be happy to contribute a PR for this!

**Anything else you would like to add:**

For reference, here are the supported tasks as of now:
https://github.com/kserve/kserve/blob/f6782431cd0751b3dc504fd51d56da75e6e12027/python/huggingfaceserver/huggingfaceserver/task.py#L18-L30

**Links to the design documents:**
[Optional, start with the short-form [RFC template](https://docs.google.com/document/d/1UcBeLfZ_JMGpVrPJmYtEIVH_9Y4U3AEKQdq_IKuOMrU) to outline your ideas and get early feedback.]
[Required, use the longer-form [design doc template](https://docs.google.com/document/d/1Mtoui_PP2a9N59NjYHnsvrdJ8t2iKFwIJAx1zRO_I1c) to specify and discuss your design in more detail]

### PR DESCRIPTION
**Pull request by kevinmingtarja**
**What this PR does / why we need it**:
Currently, the hugging face server does not support text embedding task. Text embedding seems to be an increasingly popular task, so this would be good to add.

**Which issue(s) this PR fixes**:
Fixes #3572 

**Type of changes**
- [x] New feature (non-breaking change which adds functionality)

**Feature/Issue validation/testing**:

- [x] Added a test in `test_model.py`:

```bash
% poetry run pytest -W ignore -k test_text_embedding   
The currently activated Python version 3.12.2 is not supported by the project (>=3.9,<3.12).
Trying to find and use a compatible version. 
Using python3 (3.11.5)
==================================================================== test session starts ====================================================================
platform darwin -- Python 3.11.5, pytest-7.4.4, pluggy-1.4.0
rootdir: /Users/kevin/git/kserve/python/huggingfaceserver
plugins: asyncio-0.20.3, anyio-4.3.0, cov-4.1.0, httpx-0.28.0
asyncio: mode=Mode.STRICT
collected 20 items / 19 deselected / 1 selected                                                                                                             

huggingfaceserver/test_model.py .                                                                                                                     [100%]

============================================================= 1 passed, 19 deselected in 6.65s ==============================================================
```

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
Support text embedding task in Hugging Face model server
```

**Re-running failed tests**

- `/rerun-all` - rerun all failed workflows.
- `/rerun-workflow <workflow name>` - rerun a specific failed workflow. Only one workflow name can be specified. Multiple /rerun-workflow commands are allowed per comment.

### CODE REVIEW

File: python/huggingfaceserver/huggingfaceserver/encoder_model.py [hunk 1]
```diff
@@ -242,7 +247,9 @@ async def predict(
             input_batch = input_batch.to(self._device)
             try:
                 with torch.no_grad():
-                    outputs = self._model(**input_batch).logits
+                    outputs = self._model(**input_batch)
+                    if self.task != MLTask.text_embedding.value:
```
Comment (yuzisun) [inline python/huggingfaceserver/huggingfaceserver/encoder_model.py]: good to add a comment why embedding task does not need logits

File: python/huggingfaceserver/huggingfaceserver/encoder_model.py [hunk 2]
```diff
@@ -299,6 +306,15 @@ def postprocess(
                     predictions = torch.argmax(output, dim=2)
                     inferences.append(predictions.tolist())
             return get_predict_response(request, inferences, self.name)
+        elif self.task == MLTask.text_embedding:
+            # Perform pooling
+            outputs = _mean_pooling(outputs, context["attention_mask"])
```
Comment (yuzisun) [inline python/huggingfaceserver/huggingfaceserver/encoder_model.py:311]: Do we want to depend on this private function from hugging face transformer?

File: python/huggingfaceserver/huggingfaceserver/encoder_model.py (hunk 1 above)
Comment (kevinmingtarja) [reply python/huggingfaceserver/huggingfaceserver/encoder_model.py]: got it, i refactored the code a bit to make this part about the `outputs` clearer!

**Comment by sivanantha321** [comment]
@kevinmingtarja Can you add a end to end test as well ?

**Comment by johnugeorge** [comment]
Can you provide sample input and output response

**Comment by kevinmingtarja** [comment]
> @kevinmingtarja Can you add a end to end test as well ?

Yeah for sure, let me add it.

> Can you provide sample input and output response

Run cmd: `poetry run python -m huggingfaceserver --model_id=mixedbread-ai/mxbai-embed-large-v1 --model_name=embedding --task=text_embedding`

Example:
```bash
% curl -H "Content-Type: application/json" "localhost:8080/v1/models/embedding:predict" -d "{\"instances\": [\"test\"]}" | jq
{
  "predictions": [
    [
      0.011772512458264828,
      0.017266105860471725,
      0.01167663186788559,
      0.0310803409665823,
      -0.03724151849746704,
      0.01217631995677948,
      0.009809614159166813,
      ...
      ...
      7.307787018362433e-05,
      -0.004099625162780285,
      -0.026937389746308327
    ]
  ]
}
```

Note that the output is different from if you try it on the Inference API in Hugging Face's website: https://huggingface.co/mixedbread-ai/mxbai-embed-large-v1?text=test

Screen Shot 2024-06-19 at 16 09 38

And that is because i added a `postprocess` step (mean pooling), whereas the Hugging Face API simply returns the output without any postprocessing.
https://github.com/kserve/kserve/blob/61f5a4db0bab63202ad365b66b56975a1974223f/python/huggingfaceserver/huggingfaceserver/encoder_model.py#L312-L320

File: python/huggingfaceserver/huggingfaceserver/encoder_model.py [hunk 3]
```diff
@@ -299,6 +309,15 @@ def postprocess(
                     predictions = torch.argmax(output, dim=2)
                     inferences.append(predictions.tolist())
             return get_predict_response(request, inferences, self.name)
+        elif self.task == MLTask.text_embedding:
```
Comment (johnugeorge) [inline python/huggingfaceserver/huggingfaceserver/encoder_model.py:319]: Can you explain why this post processing is required?

Comment (kevinmingtarja) [reply python/huggingfaceserver/huggingfaceserver/encoder_model.py:319]: Yeah absolutely! I first saw this on https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2#usage-huggingface-transformers.

Basically pooling is a way to obtain a sentence embedding from the embeddings of the individual word/token that make up the sentence, by averaging them.

Here's an example to illustrate that:
```python
from transformers import AutoTokenizer, AutoModel
import torch
import torch.nn.functional as F

# Mean Pooling - Take attention mask into account for correct averaging
def mean_pooling(model_output, attention_mask):
    token_embeddings = model_output.last_hidden_state
    input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
    return torch.sum(token_embeddings * input_mask_expanded, 1) / torch.clamp(input_mask_expanded.sum(1), min=1e-9)


# Sentences we want sentence embeddings for
sentences = ['Hello there, friend!', 'Good morning.']

# Load model from HuggingFace Hub
tokenizer = AutoTokenizer.from_pretrained('sentence-transformers/all-MiniLM-L6-v2')
model = AutoModel.from_pretrained('sentence-transformers/all-MiniLM-L6-v2')

# Tokenize sentences
encoded_input = tokenizer(sentences, padding=True, truncation=True, add_special_tokens=True, return_tensors='pt')
for sentence in sentences:
    print(tokenizer.tokenize(sentence, add_special_tokens=True))

# Compute token embeddings
with torch.no_grad():
    model_output = model(**encoded_input)

print("model_output.last_hidden_state.shape: ", model_output.last_hidden_state.shape)

# Perform pooling
sentence_embeddings = mean_pooling(model_output, encoded_input['attention_mask'])
print("sentence_embeddings.shape", sentence_embeddings.shape)
```

Outputs:
```bash
['[CLS]', 'hello', 'there', ',', 'friend', '!', '[SEP]']
['[CLS]', 'good', 'morning', '.', '[SEP]']
model_output.last_hidden_state.shape:  torch.Size([2, 7, 384])
sentence_embeddings.shape torch.Size([2, 384])
```

Here, we see that the first sentence is split into 7 tokens, and so the size of the embeddings of each sentence at first is `7x384`. But we care about the sentence embeddings as a whole (not the individual words), and so the mean pooling averages them to obtain the sentence embedding, a 1D array of length 384.

And finally we normalize the embeddings to ensure that the vector has a unit length (omitted from the above).

Mean pooling is a good default for models that have not been fine-tuned, as it assumes that all parts of the sentence are represented equally. Whereas `[CLS]` pooling is said to be more effective for fine-tuned models.

References:
- https://osanseviero.github.io/hackerllama/blog/posts/sentence_embeddings/
- https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/models/Pooling.py

File: test/e2e/data/text_embedding_input_v2.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
+{
+  "inputs": [
+    {
+      "name": "input-0",
+      "datatype": "BYTES",
+      "shape": [],
+      "data": ["This is an example sentence."]
+    }
+  ]
+}
```
Comment (yuzisun) [inline test/e2e/data/text_embedding_input_v2.json:10]: should we consider implementing using the openai embedding schema?
https://platform.openai.com/docs/guides/embeddings/what-are-embeddings

Comment (yuzisun) [reply test/e2e/data/text_embedding_input_v2.json:10]: We have already generated the type in kserve
https://github.com/kserve/kserve/blob/master/python/kserve/kserve/protocol/rest/openai/types/openapi.py#L769

**Review by yuzisun — COMMENTED** [review]
Will handle the inference protocol in another PR, this is good to merge

/lgtm
/approve

**Review by yuzisun — APPROVED** [review]
